src/wallet/wallet.cpp

```diff
@@ -1385,19 +1385,26 @@ bool CWallet::IsWalletFlagSet(uint64_t flag) const
     return (m_wallet_flags & flag);
 }
 
-bool CWallet::SetWalletFlags(uint64_t overwriteFlags, bool memonly)
+bool CWallet::LoadWalletFlags(uint64_t flags)
 {
     LOCK(cs_wallet);
-    m_wallet_flags = overwriteFlags;
-    if (((overwriteFlags & KNOWN_WALLET_FLAGS) >> 32) ^ (overwriteFlags >> 32)) {
+    m_wallet_flags = flags;
+    if (((flags & KNOWN_WALLET_FLAGS) >> 32) ^ (flags >> 32)) {
         // contains unknown non-tolerable wallet flags
         return false;
     }
-    if (!memonly && !WalletBatch(*database).WriteWalletFlags(m_wallet_flags)) {
+
+    return true;
+}
+
+bool CWallet::AddWalletFlags(uint64_t flags)
+{
+    LOCK(cs_wallet);
+    if (!WalletBatch(*database).WriteWalletFlags(flags)) {
```

Comment:
In the existing code, on Add, the `contains unknown non-tolerable wallet flags` test and bail will occur first and prevent writing the database. With this change, the values are written, and the failure is only identified after, which seems improper.

Reply:
Perhaps add a test for this case?

Reply:
That failure is only relevant when loading a wallet from disk. `AddWalletFlags` is only used internally in wallet creation; we would never add flags that we don't know.

Reply:
I agree that this seems less safe than the existing code. Is it possible to maintain the current order of checking first and then writing to disk?

Reply:
> I agree that this seems less safe than the existing code.

In current code m_wallet_flags can be updated even if database write fails leading to inconsistency between disk and memory state and possibility for data inconsistent with flags to subseqently be written to wallet. New code writing to disk first, memory second, prevents this type of issue and would seem to be more safe. What's the scenario where it's less safe?

Reply:
If this function is called with an unknown wallet flag, it'll be written to disk and then won't be applied to the in-memory wallet (and won't be able to be loaded later), which is also a data inconsistency.

The unknown flag check was added to protect when loading a wallet from a newer version on an older version, so I think your concern is more valid. I don't see any harm in a sanity assert that we don't try to add an unknown flag to the wallet before writing to disk.

Reply:
I've added an assert in `AddWalletFlags` to handle that case.